File: code-examples/Python/mvn_plots.py
#!/usr/bin/env python

#
# Author: Fangtong Liu
# Date: 03/22/2020
#

import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import multivariate_normal, norm


# MVN
mu = np.array([0, 0.5])  # mean
Sigma = np.array([[.8, .3], [.3, 1]])  # covariance
# compute PDF
x1 = np.arange(-3, 3.1, .1)
x2 = np.arange(-3, 4.1, .1)
X1, X2 = np.meshgrid(x1, x2)
X1X2 = np.transpose(np.array([X1.reshape(-1), X2.reshape(-1)]))
F = multivariate_normal.pdf(X1X2, mu, Sigma)
F = F.reshape((np.size(x2), np.size(x1)))

fig = plt.figure()
ax = fig.gca(projection='3d')
surf = ax.plot_surface(X1, X2, F, cmap='viridis', linewidth=0, antialiased=False)
ax.set_xlim(-3, 3)
ax.set_ylim(-3, 4)
ax.set_zlim(0, .2)
ax.set_xlabel(r'$x_1$')
ax.set_ylabel(r'$x_2$')
ax.set_zlabel(r'PDF')
# fig.colorbar(surf, shrink=0.25, aspect=6)
fig.savefig("mvn_surfave.png")
plt.show()

# marginals
f1 = norm.pdf(x1, mu[0], Sigma[0][0])
f2 = norm.pdf(x2, mu[1], Sigma[1][1])
# x_1 conditioned on x_2
mu_12 = mu[0] + Sigma[0][1] * (1 / Sigma[1][1]) * (0.9 - 0.5)  # conditional mean
sigma_12 = Sigma[0][0] - Sigma[0][1] * (1 / Sigma[1][1]) * Sigma[1][0]  # conditional variance
f12 = norm.pdf(x1, mu_12, sigma_12)
# marginal plot
fig = plt.figure()
line1, = plt.plot(x1, f1)
line2, = plt.plot(x2, f2, linestyle='--')
line3, = plt.plot(x1, f12, linestyle='-.')
plt.grid(True)
plt.xlabel(r'$x$')
plt.ylabel(r'PDF')
plt.legend([line1, line2, line3], [r'$p(x_1)$', r'$p(x_2)$', r'$p(x_1 | x_2 = 0.9)$'], loc='upper right')
plt.xlim(-4, 4)
plt.ylim(0, 0.6)
plt.text(-3.95, .57, r'$p(x_1) = \mathcal{N}(0, 0.8)$', size=12)
plt.text(-3.95, .54, r'$p(x_2) = \mathcal{N}(0.5, 1)$', size=12)
plt.text(-3.95, .51, r'$p(x_1 | x_2 = 0.9) = \mathcal{N}(0.12, 0.71)$', size=12)
fig.savefig("mvn_marginals_cond.pdf")
plt.show()

fig, ax = plt.subplots()
cp = ax.contourf(x1, x2, F, corner_mask="True", levels=30)
ax.set_xlabel(r'$x_1$')
ax.set_ylabel(r'$x_2$')
fig.savefig("mvn_top.png")
plt.show()

fig, ax = plt.subplots()
cp = ax.contour(x1, x2, F, levels=10)
ax.set_xlim(-3, 3)
ax.set_ylim(-3, 4)
ax.set_xlabel(r'$x_1$')
ax.set_ylabel(r'$x_2$')
plt.clabel(cp, inline=True, fontsize=8, colors='black')
plt.text(-2.5, 3.25, r'$\mathcal{N}([0; 0.5], [0.8, 0.3; 0.3, 1])$')
fig.savefig("mvn_contour.pdf")
plt.show()

File: code-examples/MATLAB/mvn_plots.m
%{  
    Copyright (C) 2017  Maani Ghaffari Jadidi
    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details. 
%}

clc; clear; close all

fsize = 24; % font size
set(groot, 'defaultAxesTickLabelInterpreter','latex'); 
set(groot, 'defaultLegendInterpreter','latex');

% MVN
mu = [0 0.5]; % mean
Sigma = [.8 0.3; 0.3 1]; % covariance
% compute PDF
x1 = -3:.1:3; x2 = -3:.1:4;
[X1,X2] = meshgrid(x1,x2);
F = mvnpdf([X1(:) X2(:)],mu,Sigma);
F = reshape(F,length(x2),length(x1));

figure;
surf(x1,x2,F); shading flat
axis([-3 3 -3 4 0 .2])
xlabel('$x_1$', 'Interpreter','latex'); 
ylabel('$x_2$', 'Interpreter','latex');
zlabel('PDF', 'Interpreter','latex');
set(gca,'fontsize',fsize)
set(gca,'TickLabelInterpreter','latex')
figuresize(21,21,'cm')
print -opengl -dpng -r300 mvn_surface.png

% marginals
f1 = normpdf(x1, mu(1), Sigma(1,1));
f2 = normpdf(x2, mu(2), Sigma(2,2));
% x_1 conditioned on x_2
mu_12 = mu(1) + Sigma(1,2) * (1/Sigma(2,2)) * (0.9 - 0.5); % conditional mean
sigma_12 = Sigma(1,1) - Sigma(1,2) * (1/Sigma(2,2)) * Sigma(2,1);  % conditional variance
f12 = normpdf(x1, mu_12, sigma_12);
% marginal plot
figure; hold on
h = []; % plot handle
h{1} = plot(x1, f1, 'linewidth', 2);
h{2} = plot(x2, f2, '--', 'linewidth', 2);
h{3} = plot(x1, f12, '-.', 'linewidth', 2);
axis auto, grid on
xlabel('$x$', 'Interpreter','latex'); 
ylabel('PDF', 'Interpreter','latex');
legend([h{1}, h{2}, h{3}], '$p(x_1)$', '$p(x_2)$', '$p(x_1|x_2 = 0.9)$', 'location', 'northeast')
text(-3.95, 0.58, '$p(x_1) = \mathcal{N}(0,0.8)$', 'fontsize',16, 'Interpreter','latex')
text(-3.95, 0.54, '$p(x_2) = \mathcal{N}(0.5,1)$', 'fontsize',16, 'Interpreter','latex')
text(-3.95, 0.51, '$p(x_1|x_2 = 0.9) = \mathcal{N}(0.12,0.71)$', 'fontsize',16, 'Interpreter','latex')
set(gca,'fontsize',fsize)
set(gca,'TickLabelInterpreter','latex')
figuresize(21,21,'cm')
print -painters -dpdf -r600 mvn_marginals_cond.pdf

figure;
pcolor(x1,x2,F);
shading interp, grid on
axis([-3 3 -3 4])
xlabel('$x_1$', 'Interpreter','latex'); 
ylabel('$x_2$', 'Interpreter','latex');
set(gca,'fontsize',fsize)
set(gca,'TickLabelInterpreter','latex')
figuresize(21,21,'cm')
print -opengl -dpng -r300 mvn_top.png

figure;
contour(x1,x2,F, 'ShowText','on')
axis tight, grid on
xlabel('$x_1$', 'Interpreter','latex'); 
ylabel('$x_2$', 'Interpreter','latex');
text(-2.5, 3.25, '$\mathcal{N}([0;0.5], [0.8,0.3;0.3,1])$', 'fontsize',fsize, 'Interpreter','latex')
set(gca,'fontsize',fsize)
set(gca,'TickLabelInterpreter','latex')
figuresize(21,21,'cm')
print -painters -dpdf -r600 mvn_contour.pdf
